### DISCUSSION TITLE
How to make sure Ajax re-authentication requests not to fail on CORS?

### DISCUSSION THREAD

**Discussion by zeppelinux** [keycloak/keycloak#20910] [Q&A]
Hi All,

I'm having this problem:
https://stackoverflow.com/questions/76284133/how-to-deal-with-quarkus-oidc-cookie-expiration-and-ajax-requests-cors-failure

Any idea for how to fix it or reduce the pain (forcing users to purge their cache is not a good way) would be greatly appreciated.

Also asked Primefaces community https://forum.primefaces.org/viewtopic.php?f=3&t=81158

**Comment by zeppelinux** [comment]
Cross referencing:
https://keycloak.discourse.group/t/no-access-control-allow-origin-header-is-present-on-the-requested-resource-in-keycloak-21/22539/2

**Comment by zeppelinux** [comment]
Adding @sberyozkin. I'm still looking for some pointers to resolve this.

**Comment by sberyozkin** [comment]
@zeppelinux 



Keycloak Authorization endpoint does not support cross-origin redirects made with XHR, therefore the script has to handle such redirects as described in

https://quarkus.io/guides/security-oidc-code-flow-authentication-concept#single-page-applications

Going forward it is better to open Quarkus dicussions first - and then we can decide if some help from the Keycloak team is needed or not,

Thanks

**Comment by dteleguin** [comment]
As @sberyozkin mentioned, CORS is not supported on Authorization endpoint - just wanted to add that this is by design (and not a bug). The most recent version of OAuth 2.0 Security Best Current Practice is explicit on that:

> To support browser-based clients, endpoints directly accessed by such clients including the Token Endpoint, Authorization Server Metadata Endpoint, jwks_uri Endpoint, and the Dynamic Client Registration Endpoint MAY support the use of Cross-Origin Resource Sharing (CORS, [[CORS](https://fetch.spec.whatwg.org/#http-cors-protocol)]). However, **CORS MUST NOT be supported at the Authorization Endpoint** as the client does not access this endpoint directly, instead the client redirects the user agent to it.

Simply put - if your client-side code tries to access Authorization Endpoint programmatically, then this is the issue that needs to be fixed in the long term. In the short term, you can try several workarounds:
- use Quarkus CORS filter. I'd recommend that only as a last resort as it is non-selective (will enable CORS just everywhere) and will [break Keycloak](https://github.com/quarkusio/quarkus/issues/31631) unless you add `null` to allowed origins (which could potentially pose a security risk);
- deploy a custom JAX-RS filter that would bind to Authorization endpoint only and provide CORS support for specific requests. (Ping me on Keycloak Slack if you need more help with that.)
